moshi/src/main/java/com/squareup/moshi/JsonUtf8Writer.java

```diff
@@ -273,16 +276,34 @@ private void writeDeferredName() throws IOException {
     return this;
   }
 
-  @Override public JsonWriter value(BufferedSource source) throws IOException {
+  @Override public BufferedSink valueSink() throws IOException {
     if (promoteValueToName) {
       throw new IllegalStateException(
-          "BufferedSource cannot be used as a map key in JSON at path " + getPath());
+          "BufferedSink cannot be used as a map key in JSON at path " + getPath());
     }
     writeDeferredName();
     beforeValue();
-    sink.writeAll(source);
-    pathIndices[stackSize - 1]++;
-    return this;
+    pushScope(STREAMING_VALUE);
+    return Okio.buffer(new Sink() {
+      @Override public void write(Buffer source, long byteCount) throws IOException {
+        sink.write(source, byteCount);
+      }
+
+      @Override public void close() {
+        if (peekScope() != STREAMING_VALUE) {
+          throw new AssertionError();
+        }
+        stackSize--; // Remove STREAMING_VALUE from the stack.
+        pathIndices[stackSize - 1]++;
+      }
+
+      @Override public void flush() {
```

Comment:
flush the underlying sink!

Reply:
I thought it would be weird that you could do this since you cannot do it anywhere else. I don't particularly mind hooking it up, but it didn't seem like something you should have access to.

Reply:
We could potentially do the thing where you can constrain the timeouts but you can’t relax them?

Reply:
SGTM. Is there an example of this somewhere?